src/dinit-iostream.cc

```diff
@@ -0,0 +1,759 @@
+{
+    int bits = 0;
+    if (eofbit) bits |= io_states::eofbit;
+    if (buffer_failbit) bits |= io_states::buffer_failbit;
+    if (string_failbit) bits |= io_states::string_failbit;
+    if (io_error) bits |= io_states::io_error;
+    return bits;
+}
+
+bool istream::good() noexcept
+{
+    return (current_state() == 0);
+}
+
+bool istream::eof() noexcept
+{
+    return (current_state() & io_states::eofbit);
+}
+
+bool istream::buffer_failure() noexcept
+{
+    return (current_state() & io_states::buffer_failbit);
+}
+
+bool istream::string_failure() noexcept
+{
+    return (current_state() & io_states::string_failbit);
+}
+
+int istream::io_failure() noexcept
+{
+    return io_error;
+}
+
+void istream::clear() noexcept
+{
+    eofbit = false;
+    buffer_failbit = false;
+    string_failbit = false;
+    io_error = 0;
+}
+
+int istream::getc_nx() noexcept
+{
+    if (eof() || buffer_failure() || io_failure()) {
+        // We don't care about string_failure() in getc() function.
+        return -1;
+    }
+    if (buf->get_length() == 0) {
+        int r = load_into_buf(1);
+        if (r == 0) {
+            eofbit = true;
+            return -1;
+        }
+        else if (r < 0) {
+            io_error = errno;
+            return -1;
+        }
+    }
+    char result = *buf->get_ptr(0);
+    buf->consume(1);
+    return result;
+}
+
+int istream::getc()
+{
+    int r = getc_nx();
+    if (r <= 0) {
+        if (buffer_failure()) {
+            throw iostream_internal_err("Cannot use stream's buffer: buffer is nullptr!");
+        }
+        if (eof()) {
+            throw iostream_eof("End Of File has been reached.");
+        }
+        if (io_failure()) {
+            throw iostream_system_err(io_failure());
+        }
+    }
+    return r;
+}
+
+ssize_t istream::get_line_nx(std::string &dest, char delim) noexcept
+{
+    if (!good()) {
+        return -1;
+    }
+    if (buf->get_length() == 0) {
+        int r = load_into_buf(buf->get_size());
+        if (r == 0) {
+            eofbit = true;
+            return 0;
+        }
+        if (r < 0) {
+            io_error = errno;
+            return -1;
+        }
+    }
+    dest.clear();
+
+    char *ptr = buf->get_ptr(0);
+    unsigned len = buf->get_contiguous_length(ptr);
+    char *eptr = static_cast<char*>(std::memchr(ptr, delim, len));
+
+    if (eptr == ptr) {
+        buf->consume(1);
+        return 1; // An empty line
+    }
+    while (eptr == nullptr) {
+        // If it hasn't any delim; We try to read as much as possible to reach the EOF
+        // or find a "delim" character.
+        try {
+            // dest.append() may throw out_of_range, length_error or bad_alloc
+            dest.append(ptr, len);
+        }
+        catch (const std::exception &e) {
+            string_failbit = true;
+            return -1;
+        }
+        buf->consume(len);
+
+        int r = load_into_buf(buf->get_size());
+        if (r == 0) {
```

Comment:
There is a bug here I think.

The buffer is circular. Calling `load_into_buf` will return 0 if end-of-file is reached, but that doesn't mean the buffer is empty, it's still necessary to check the remaining half of the buffer.

Also, the buffer should be fully emptied before attempting to read more from the file.

Currently the logic is:

1. process (half) the buffer and consume it
2. load more into the buffer
3. (repeat from 1)

It should be:

1. process (half) the buffer and consume it
2. if buffer is not empty, process remainder and consume it
3. load more into the buffer
4. (repeat from 1)

Reply:
```c++
        if (buf->get_length() == 0) {
            int r = load_into_buf(buf->get_size());
            if (r == 0) {
                return dest.size();
            }
            if (r < 0) {
                io_error = errno;
                return -1;
            }
        }
```

Reply:
I don't understand why you posted this as a comment, is that how you are proposing to fix it? If so just make the change

Reply:
I did post it because I want to make sure I did understand your opinion correctly.

Reply:
Yes, what you have now looks correct to me